feat(ui): build machine power configuration form

Caleb-Ellis opened the pull request:
## Done

- Built "Power configuration" section of the config tab
- Changed links to point to the React version of the config tab
- Updated `FormikForm`'s `onCancel` prop so you can give it the entire formik context as an argument and access things like `initialValues` and `resetForm`. I've used it in the MachineForm and PowerForm to reset the form when "Cancel" is pressed.

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Create a machine manually (i.e. not in a pod) and go to its details page
- Click "Configuration" and check that it takes you to the React configuration screen
- Check that the power details are initialised correctly
- Click the "Edit" button next to "Power configuration"
- Check that the select is enabled and you can change the power type and power parameters
- Change a few values then click "Cancel" and check that the values are reverted to the originals
- Enter an invalid value (e.g "1" for an ip address) and submit and check that en error is displayed
- Check that you can successfully update the power config with valid values
- Go to the config page for a machine that's in a pod
- Check that the select is always disabled, and you only see fields that are scoped to individual nodes

## Fixes

Fixes canonical-web-and-design/MAAS-squad#2223

huwshimi commented:
@Caleb-Ellis a few QA things.

Non-pod machine config:
- If you change the power type then press cancel it doesn't revert the form to the old power type.
- If you enter a value for e.g. ip address and then change the power type then if there is another field for ip address in the new form it has the value you previously entered, is that intentional? At first I wasn't sure if it had prefilled the field some other way.
- If you have a form error it shows in the power form and the machine config form above it.
- If you get an error and then cancel it doesn't clear the error.
- I got an error for a bad ip address format that was really confusing. Could we do IP validation on the client or is it not possible because the fields are generated?
<img width="1076" alt="Screen Shot 2021-02-15 at 11 16 56 am" src="https://user-images.githubusercontent.com/361637/107893674-a2a2b000-6f80-11eb-92c6-e3e7ac6daf9a.png">

Pod machine config:
- There's a warning about changing config for fields that aren't shown. Do we have any control over that warning?
<img width="538" alt="Screen Shot 2021-02-15 at 11 19 27 am" src="https://user-images.githubusercontent.com/361637/107893543-fa8ce700-6f7f-11eb-8737-37a6629dc8eb.png">

Caleb-Ellis commented:
@huwshimi thanks for the review! I thought breaking this form down into several smaller issues would've ironed out the kinks, but clearly not!

>If you change the power type then press cancel it doesn't revert the form to the old power type.

Fixed. I've updated `FormikForm`'s `onCancel` prop so you can give it the entire formik context as an argument and access things like `initialValues` and `resetForm`.

>If you enter a value for e.g. ip address and then change the power type then if there is another field for ip address in the new form it has the value you previously entered, is that intentional? At first I wasn't sure if it had prefilled the field some other way.

Fixed. In `PowerTypeFields` I've updated the select to set the power type defaults on change and to reset touched/errors so they don't carry over.

>If you have a form error it shows in the power form and the machine config form above it.

I'm not sure how to fix this one. I've changed it so that errors only show if the machine is in editing state, but they're both `machine.update` forms with different fields so they both have the same errors associated with them in state. Any ideas? Maybe something to address in a follow up?

>If you get an error and then cancel it doesn't clear the error.

Fixed. Added `cleanup` to the `onCancel` function.

>I got an error for a bad ip address format that was really confusing. Could we do IP validation on the client or is it not possible because the fields are generated?

I think if we want to do IP validation on the client the power field from the api should specify as such, otherwise there's the potential for the validation to be wrong. Right now we only know whether a field is a `string`, `password` or `choice`.

>There's a warning about changing config for fields that aren't shown. Do we have any control over that warning?

I actually just copied over the logic from the angular version, but I think the warning might be unnecessary. The fields that are displayed are limited to `PowerFieldScope.NODE` if the machine is in a pod, so it shouldn't be possible to change anything that'll affect other nodes. Removing for now.

Caleb-Ellis commented on ui/src/app/base/components/FormikForm/FormikForm.tsx:
```diff
@@ -23,7 +23,7 @@ export type Props<V, E = FormErrors> = {
   onSubmit: (values?: V, formikHelpers?: FormikHelpers<V>) => void;
   savedRedirect?: string;
-  validationSchema?: SchemaOf<V>;
+  validationSchema?: AnySchema;
```
This was causing a typescript error for any `FormikForm` that was given a generic

huwshimi replied:
I think the previous type was correct, but looks like you have to do a little bit more setup to define the type of the Schema:

```
  const PowerFormSchema: SchemaOf<PowerFormValues> = Yup.object()
    .shape({
      powerParameters: Yup.object().shape(powerParametersSchema),
      powerType: Yup.string().required("Power type is required"),
    })
    .defined();
```
From: https://github.com/jquense/yup/blob/master/docs/typescript.md

Caleb-Ellis replied:
Aaaaaaahhhhh I missed the `defined()` bit when I was looking at that earlier. Nice, thanks, I'll revert the change

huwshimi commented:
@Caleb-Ellis this looks mostly good now except if you make some changes and save then make some more changes and press cancel the changes go back to the original version before the first save.

huwshimi commented on ui/src/app/base/components/PowerTypeFields/PowerTypeFields.tsx:
```diff
@@ -101,12 +117,30 @@ export const PowerTypeFields = <F extends Record<string, unknown>>({
               value: powerType.name,
             })),
           ]}
+          onChange={(e: React.ChangeEvent<HTMLSelectElement>) => {
```
I think what this does is loop through and reset any fields that are for the selected power type. Should it also reset the values for the old power type (or maybe just everything) as couldn't it leave values in the formik state for the old fields? Not sure exactly how that works.

Either way, it would probably also good to leave a comment here explaining why this is necessary.

Caleb-Ellis replied:
It shouldn't be necessary to reset any other fields than from the selected power type because the power parameters are trimmed before the websocket call is made. Resetting everything won't work (at least using `resetForm()`) because it'll end up resetting non-power-related form fields as well.

I'll leave a comment explaining what this is for (and add a test)

huwshimi commented on ui/src/app/machines/views/MachineDetails/MachineConfiguration/MachineForm/MachineForm.tsx:
```diff
@@ -84,7 +84,7 @@ const MachineForm = ({ systemId }: Props): JSX.Element | null => {
           cleanup={cleanup}
           editable={editing}
-          errors={errors}
+          errors={editing ? errors : undefined}
```
Does this need to be `undefined` here? If so maybe leave a comment (or maybe see if it can be fixed so it doesn't have to be `undefined`) otherwise this should probably use `null`.

Caleb-Ellis replied:
Changed to `null`

huwshimi commented on ui/src/app/machines/views/MachineDetails/MachineConfiguration/PowerForm/PowerForm.test.tsx:
```diff
@@ -0,0 +1,142 @@
+
+    wrapper.find("button[data-test='edit-power-config']").simulate("click");
+
+    expect(wrapper.find("FormikForm").prop("editable")).toBe(true);
```
Looks

like

this

could

probably

have

less

empty

lines.

Caleb-Ellis replied:
But





why?

huwshimi commented on ui/src/app/machines/views/MachineDetails/MachineConfiguration/PowerForm/PowerForm.tsx:
```diff
@@ -0,0 +1,151 @@
+          buttons={FormCardButtons}
+          cleanup={cleanup}
+          editable={editing}
+          errors={editing ? errors : undefined}
```
`null` here as well?

huwshimi commented on ui/src/app/machines/views/MachineDetails/MachineConfiguration/PowerForm/PowerForm.test.tsx:
```diff
@@ -0,0 +1,142 @@
+      type: "machine/update",
+    });
+  });
+});
```
Might also be good to add a couple of tests covering the QA issues I found (resetting fields etc.) so that we catch any regressions.

Caleb-Ellis replied:
Good call. I've added tests for:
- Power fields being set to defaults when the power type is changed
- MachineForm/PowerForm only showing errors in editing state
- MachineForm/PowerForm being reset to initial values when for is cancelled
- MachineForm/PowerForm dispatching a machine cleanup action on cancel

huwshimi reviewed (changes requested):
A few changes required here sorry.

Caleb-Ellis commented:
@huwshimi Alright I think this is the one!

huwshimi reviewed (approved):
Looks good.
